File: android/src/main/java/com/evva/xesar/abrevva/plugins/capacitor/AbrevvaPluginCodingStation.kt
package com.evva.xesar.abrevva.plugins.capacitor

import android.Manifest
import android.content.Intent
import com.evva.xesar.abrevva.auth.AuthManager
import com.evva.xesar.abrevva.cs.CodingStation
import com.evva.xesar.abrevva.mqtt.MqttConnectionOptions
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.annotation.Permission
import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import java.net.URL

@CapacitorPlugin(
    name = "AbrevvaPluginCodingStation", permissions = [
        Permission(strings = [Manifest.permission.NFC], alias = "NFC"),
    ]
)
class AbrevvaPluginCodingStation : Plugin() {

    private var connectionOptions: MqttConnectionOptions? = null
    private lateinit var cs: CodingStation

    override fun load() {
        super.load()
        cs = CodingStation(context)
    }

    @PluginMethod
    fun register(call: PluginCall) {
        val url = call.getString("url")
        val clientId = call.getString("clientId")
        val username = call.getString("username")
        val password = call.getString("password")

        if (url == null || clientId == null || username == null || password == null) {
            return call.reject("register(): invalid params")
        }
        try {
            connectionOptions = AuthManager.getMqttConfigForXS(URL(url), clientId, username, password)
            call.resolve()
        } catch (e: Exception) {
            call.reject("register(): failed to authenticate: $e")
        }
    }

    @PluginMethod
    @OptIn(DelicateCoroutinesApi::class)
    fun connect(call: PluginCall) {
        GlobalScope.launch {
            connectionOptions?.let {
                try {
                    cs.connect(it)
                    call.resolve()
                } catch (e: Exception) {
                    call.reject("connect(): failed to connect: $e")
                }
            } ?: run {
                call.reject("connect(): not authenticated")
            }
        }
    }

    @PluginMethod
    fun disconnect(call: PluginCall) {
        cs.disconnect()
        call.resolve()
    }

    @PluginMethod
    @OptIn(DelicateCoroutinesApi::class)
    fun write(call: PluginCall) {
        GlobalScope.launch {
            try {
                val success = cs.startTagReader(activity, 10000L)
                if (success) call.resolve()
                else call.reject("write(): write cancelled")
            } catch (e: Exception) {
                call.reject("write(): failed to write: $e")
            }
        }
    }

    override fun handleOnNewIntent(intent: Intent?) {
        super.handleOnNewIntent(intent).also { cs.onHandleIntent(intent) }
    }
}

File: ios/Plugin/cs/AbrevvaPluginCodingStation.swift
import Capacitor
import Foundation
import AbrevvaSDK

@objc(AbrevvaPluginCodingStation)
public class AbrevvaPluginCodingStation: CAPPlugin {

    private var connectionOptions: MqttConnectionOptions?
    private let cs = CodingStation()

    @objc
    func register(_ call: CAPPluginCall) {
        let url = call.getString("url")
        let clientId = call.getString("clientId")
        let username = call.getString("username")
        let password = call.getString("password")

        if let _url = url, let _clientId = clientId, let _username = username, let _password = password {
            Task {
                do {
                    connectionOptions = try await AuthManager.getMqttConfigForXS(url: URL(string: _url)!, clientId: _clientId, username: _username, password: _password)
                    call.resolve()
                } catch {
                    return call.reject("register(): failed to authenticate: \(error)")
                }
            }
        } else {
            call.reject("register(): invalid params")
        }
    }

    @objc
    func connect(_ call: CAPPluginCall) {
        if let connectionOptions = connectionOptions {
            Task {
                do {
                    try await cs.connect(connectionOptions)
                    call.resolve()
                } catch {
                    call.reject("connect(): failed to connect: \(error)")
                }
            }
        } else {
            call.reject("connect(): not authenticated")
        }
    }

    @objc
    func disconnect(_ call: CAPPluginCall) {
        cs.disconnect()
        call.resolve()
    }

    @objc
    func write(_ call: CAPPluginCall) {
        Task {
            do {
                try await cs.write()
                call.resolve()
            } catch {
                call.reject("write(): failed to write: \(error)")
            }
        }
    }
}
